src/categorical_algebra/Automorphisms.jl

```diff
@@ -0,0 +1,306 @@
+module Automorphisms
+export canonical_hash, canonical_iso, autos, apply_automorphism, color_refine, pseudo_cset, pseudo_cset_inv
+using ..CSets
+using ...Theories
+using ...Theories: adom, attr, data, attr, adom, acodom, AttrDesc
+using Catlab.Present
+
+# Color assigned to each elem of each compoennt
+CDict = Dict{Symbol, Vector{Int}}
+
+"""
+To compute automorphisms of Attributed CSets, we create a pseudo CSet which has
+components for each data type.
+"""
+function pseudo_cset(g::ACSet{CD,AD})::Tuple{CSet, Vector{Vector{Any}}} where {CD, AD}
+    tabs, arrs, src, tgt = collect(ob(CD)), collect(hom(CD)), dom(CD), codom(CD)
+    dtabs, darrs, dsrc, dtgt = collect(data(AD)), collect(attr(AD)), adom(AD), acodom(AD)
+    pres = Presentation(FreeSchema)
+    xobs = [Ob(FreeSchema,t) for t in vcat(tabs,dtabs)]
+    n = length(tabs)
+    for x in xobs add_generator!(pres, x) end
+    for (arr, s, t) in zip(map(collect, [arrs, src, tgt])...)
+        add_generator!(pres, Hom(arr, xobs[s], xobs[t]))
+    end
+    attrvals = [Set() for _ in 1:length(dtabs)]
+    for (arr, s, t) in zip(darrs, dsrc, dtgt)
+        add_generator!(pres, Hom(arr, xobs[s], xobs[t+n]))
+        union!(attrvals[t], Set(g[arr]))
+    end
+    attrvals = Vector{Any}[sort(collect(x)) for x in attrvals]
+    ctype =  ACSetType(pres, index=vcat(arrs,darrs))
+    res = ctype()
+    for t in tabs add_parts!(res, t, nparts(g,t)) end
+    for (i,t) in enumerate(dtabs) add_parts!(res, t, length(attrvals[i])) end
+    for a in arrs set_subpart!(res, a, g[a]) end
+    for (a,t) in zip(darrs, dtgt)
+        fks = [findfirst(==(v), attrvals[t]) for v in g[a]]
+        set_subpart!(res, a, fks)
+    end
+    return res, attrvals
+end
+
+"""Inverse of pseudo_cset"""
+function pseudo_cset_inv(g::CSet, orig::ACSet{CD,AD}, attrvals::Vector{Vector{Any}})::ACSet{CD,AD} where {CD,AD}
+    orig = deepcopy(orig)
+    arrs = hom(CD)
+    darrs, dtgt = attr(AD), acodom(AD)
+    for arr in arrs
+        set_subpart!(orig, arr, g[arr])
+    end
+    for (darr,tgt) in zip(darrs, dtgt)
+        set_subpart!(orig, darr, attrvals[tgt][g[darr]])
+    end
+    return orig
+end
+
+# The maximum color of an empty color list is 0
+function max0(x::Vector{Int})::Int
+    return isempty(x) ? 0 : maximum(x)
+end
+
+function check_auto(x::CDict)::Nothing
+    for perm in values(x)
+        @assert length(perm) == length(Set(perm))
+    end
+end
+
+"""Apply a coloring to a Cset to get an isomorphic cset"""
+function apply_automorphism(c::CSet,d::CDict)::CSet
+    check_auto(d)
```

Comment:
if you make `check_auto` into a predicate, than this line should be 

```julia
check_auto || error("d is not auto")
```
but with an informative error message. This will ensure that when the check auto predicate fails, the user sees a meaningful error.